File: kor-common/src/test/java/com/sefford/kor/repositories/ExpirationRepositoryTest.java
/*
 * Copyright (C) 2017 Saúl Díaz
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.sefford.kor.repositories;

import com.sefford.kor.repositories.interfaces.ExpirationPolicy;
import com.sefford.kor.repositories.interfaces.Repository;
import com.sefford.kor.repositories.utils.TestElement;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;

import java.util.Arrays;
import java.util.Collection;
import java.util.HashMap;

import static org.hamcrest.core.Is.is;
import static org.junit.Assert.*;
import static org.mockito.Mockito.*;
import static org.mockito.MockitoAnnotations.initMocks;

/**
 * @author Saul Diaz <sefford@gmail.com>
 */
public class ExpirationRepositoryTest {

    Repository<Integer, TestElement> repository;
    Repository<Integer, TestElement> expirationRepo;

    @Mock
    ExpirationPolicy<Integer> policy;

    @Before
    public void setUp() throws Exception {
        initMocks(this);

        repository = new MemoryDataSource<>(new HashMap<Integer, TestElement>());
        expirationRepo = new ExpirationRepository<>(repository, policy);
    }

    @Test
    public void testCorrectClearing() throws Exception {
        expirationRepo.save(new TestElement(0));
        expirationRepo.save(new TestElement(1));
        expirationRepo.save(new TestElement(2));

        expirationRepo.clear();

        verify(policy, times(1)).clear();
        assertFalse(repository.contains(0));
        assertFalse(repository.contains(1));
        assertFalse(repository.contains(2));
    }

    @Test
    public void testContainsWhenExpired() throws Exception {
        expirationRepo.save(new TestElement(0));
        when(policy.isExpired(0)).thenReturn(Boolean.TRUE);

        assertFalse(expirationRepo.contains(0));
        verify(policy, times(1)).notifyDeleted(0);
    }

    @Test
    public void testContainsWhenNotExpiredAndInInnerRepo() throws Exception {
        expirationRepo.save(new TestElement(0));
        when(policy.isExpired(0)).thenReturn(Boolean.FALSE);

        assertTrue(expirationRepo.contains(0));
        verify(policy, never()).notifyDeleted(0);
    }

    @Test
    public void testContainsWhenNotExpiredAndNotInInnerRepo() throws Exception {
        when(policy.isExpired(0)).thenReturn(Boolean.FALSE);

        assertFalse(expirationRepo.contains(0));

        verify(policy, atLeastOnce()).notifyDeleted(0);
    }

    @Test
    public void testDeletion() throws Exception {
        expirationRepo.save(new TestElement(0));

        expirationRepo.delete(0, null);

        verify(policy, times(1)).notifyDeleted(0);
        assertFalse(repository.contains(0));
    }

    @Test
    public void testRetrievalWhenNotExpired() throws Exception {
        expirationRepo.save(new TestElement(0));
        when(policy.isExpired(0)).thenReturn(Boolean.FALSE);

        TestElement testElement = expirationRepo.get(0);

        assertNotNull(testElement);
        assertThat(testElement.getId(), is(0));
    }

    @Test
    public void testRetrievalWhenExpired() throws Exception {
        expirationRepo.save(new TestElement(0));
        when(policy.isExpired(0)).thenReturn(Boolean.TRUE);

        TestElement testElement = expirationRepo.get(0);

        assertNull(testElement);
    }

    @Test
    public void testGetPartialRetrievalWithExpiredElements() throws Exception {
        expirationRepo.save(new TestElement(0));
        expirationRepo.save(new TestElement(1));
        expirationRepo.save(new TestElement(2));

        when(policy.isExpired(1)).thenReturn(Boolean.TRUE);

        final Collection<TestElement> elements = expirationRepo.getAll(Arrays.asList(0, 1, 2));

        assertTrue(elements.contains(new TestElement(0)));
        assertTrue(elements.contains(new TestElement(2)));
    }

    @Test
    public void testFullRetrievalWithExpiredElements() throws Exception {
        expirationRepo.save(new TestElement(0));
        expirationRepo.save(new TestElement(1));
        expirationRepo.save(new TestElement(2));

        when(policy.isExpired(1)).thenReturn(Boolean.TRUE);

        final Collection<TestElement> elements = expirationRepo.getAll();

        assertTrue(elements.contains(new TestElement(0)));
        assertTrue(elements.contains(new TestElement(2)));
    }

    @Test
    public void testSaving() throws Exception {
        expirationRepo.save(new TestElement(0));

        verify(policy, times(1)).notifyCreated(0);
        assertTrue(expirationRepo.contains(0));
    }

    @Test
    public void testAvailabilityWhenPolicyIsNull() throws Exception {
        expirationRepo = new ExpirationRepository<>(repository, null);

        assertFalse(expirationRepo.isAvailable());
    }

    @Test
    public void testAvailabilityWhenRepositoryIsNull() throws Exception {
        expirationRepo = new ExpirationRepository<>(null, policy);

        assertFalse(expirationRepo.isAvailable());
    }

    @Test
    public void testAvailabilityWhenTheRepoIsWorking() throws Exception {
        assertTrue(expirationRepo.isAvailable());
    }
}

File: modules/kor-repositories/src/test/kotlin/com/sefford/kor/repositories/ExpirationRepositoryTest.kt
/*
 * Copyright (C) 2017 Saúl Díaz
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.sefford.kor.repositories

import com.nhaarman.mockito_kotlin.whenever
import com.sefford.kor.repositories.components.ExpirationPolicy
import com.sefford.kor.repositories.components.Repository
import com.sefford.kor.repositories.components.RepositoryError
import com.sefford.kor.test.TestElement
import org.hamcrest.core.Is.`is`
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.MockitoAnnotations.initMocks
import java.util.*

/**
 * @author Saul Diaz <sefford@gmail.com>
 */
class ExpirationRepositoryTest {

    lateinit var innerRepo: Repository<Int, TestElement>
    lateinit var repository: ExpirationRepository<Int, TestElement>

    @Mock
    lateinit var policy: ExpirationPolicy<Int>

    @Before
    fun setUp() {
        initMocks(this)

        innerRepo = MemoryDataSource(HashMap())
        repository = ExpirationRepository(innerRepo, policy)
    }

    @Test
    fun `clear should properly clear the repository`() {
        repository.save(TestElement(0))
        repository.save(TestElement(1))
        repository.save(TestElement(2))

        repository.clear()

        verify<ExpirationPolicy<Int>>(policy, times(1)).clear()
        assertFalse(innerRepo.contains(0))
        assertFalse(innerRepo.contains(1))
        assertFalse(innerRepo.contains(2))
    }

    @Test
    fun `contains should properly report false and delete when the element has expired`() {
        repository.save(TestElement(0))
        whenever(policy.isExpired(0)).thenReturn(true)

        assertFalse(repository.contains(0))
        verify<ExpirationPolicy<Int>>(policy, times(1)).notifyDeleted(0)
    }

    @Test
    fun `contains should properly report true when the element has not expired`() {
        repository.save(TestElement(0))
        whenever(policy.isExpired(0)).thenReturn(false)

        assertTrue(repository.contains(0))
        verify<ExpirationPolicy<Int>>(policy, never()).notifyDeleted(0)
    }

    @Test
    fun `contains should properly report false and delete when the element has not expired but it has magically been deleted from the repo`() {
        whenever(policy.isExpired(0)).thenReturn(false)

        assertFalse(repository.contains(0))

        verify<ExpirationPolicy<Int>>(policy, atLeastOnce()).notifyDeleted(0)
    }

    @Test
    fun `delete should properly update the status and delete the inner repo`() {
        repository.save(TestElement(0))

        repository.delete(0)

        verify(policy, times(1)).notifyDeleted(0)
        assertFalse(innerRepo.contains(0))
    }

    @Test
    fun `should delete in batch via varargs`() {
        repository.save(TestElement(0), TestElement(1), TestElement(2))

        repository.delete(TestElement(0), TestElement(1), TestElement(2))

        listOf(0, 1, 2).forEach { id ->
            verify(policy, times(1)).notifyDeleted(id)
            assertFalse(innerRepo.contains(id))
        }
    }

    @Test
    fun `should delete in batch via collection`() {
        repository.save(TestElement(0), TestElement(1), TestElement(2))

        repository.delete(listOf(TestElement(0), TestElement(1), TestElement(2)))

        listOf(0, 1, 2).forEach { id ->
            verify(policy, times(1)).notifyDeleted(id)
            assertFalse(innerRepo.contains(id))
        }
    }

    @Test
    fun `should delete in batch via iterator`() {
        repository.save(TestElement(0), TestElement(1), TestElement(2))

        repository.delete(listOf(TestElement(0), TestElement(1), TestElement(2)).iterator())

        listOf(0, 1, 2).forEach { id ->
            verify(policy, times(1)).notifyDeleted(id)
            assertFalse(innerRepo.contains(id))
        }
    }

    @Test
    fun `an element should be retrieved if it has not expired`() {
        repository.save(TestElement(0))
        whenever(policy.isExpired(0)).thenReturn(false)

        val testElement = repository[0].right().get()

        assertNotNull(testElement)
        assertThat(testElement.id, `is`(0))
    }

    @Test
    fun `a NotFound error should be returned if the element has expired`() {
        repository.save(TestElement(0))
        whenever(policy.isExpired(0)).thenReturn(true)

        val error = repository[0].left().get()

        assertTrue(error is RepositoryError.NotFound<*>)
    }

    @Test
    fun `only non-expired elements should be returned from varargs`() {
        repository.save(TestElement(0), TestElement(1), TestElement(2))
        whenever(policy.isExpired(1)).thenReturn(true)

        val elements = repository.get(0, 1, 2)

        assertThat(elements.size, `is`(2))
        assertTrue(elements.contains(TestElement(0)))
        assertTrue(elements.contains(TestElement(2)))
    }

    @Test
    fun `only non-expired elements should be returned from collection`() {
        repository.save(TestElement(0), TestElement(1), TestElement(2))
        whenever(policy.isExpired(1)).thenReturn(true)

        val elements = repository.get(listOf(0, 1, 2))

        assertThat(elements.size, `is`(2))
        assertTrue(elements.contains(TestElement(0)))
        assertTrue(elements.contains(TestElement(2)))
    }

    @Test
    fun `only non-expired elements should be returned from iterator`() {
        repository.save(TestElement(0), TestElement(1), TestElement(2))
        whenever(policy.isExpired(1)).thenReturn(true)

        val elements = repository.get(listOf(0, 1, 2).iterator())

        assertThat(elements.size, `is`(2))
        assertTrue(elements.contains(TestElement(0)))
        assertTrue(elements.contains(TestElement(2)))
    }

    @Test
    fun `only non-expired elements should be returned when retrieving all the elements`() {
        repository.save(TestElement(0), TestElement(1), TestElement(2))

        whenever(policy.isExpired(1)).thenReturn(true)

        val elements = repository.all

        assertThat(elements.size, `is`(2))
        assertTrue(elements.contains(TestElement(0)))
        assertTrue(elements.contains(TestElement(2)))
    }

    @Test
    fun `save should properly update the state and persis the element in the inner repo`() {
        repository.save(TestElement(0))

        verify<ExpirationPolicy<Int>>(policy, times(1)).notifyCreated(0)
        assertTrue(repository.contains(0))
    }

    @Test
    fun `elements should be able to be saved in batch via varargs`() {
        repository.save(TestElement(0), TestElement(1), TestElement(2))

        listOf(0, 1, 2).forEach { id ->
            assertThat(innerRepo.contains(id), `is`(true))
            verify(policy, times(1)).notifyCreated(id)
        }
    }

    @Test
    fun `elements should be able to be saved in batch via collection`() {
        repository.save(listOf(TestElement(0), TestElement(1), TestElement(2)))

        listOf(0, 1, 2).forEach { id ->
            assertThat(innerRepo.contains(id), `is`(true))
            verify(policy, times(1)).notifyCreated(id)
        }
    }

    @Test
    fun `elements should be able to be saved in batch via iterator`() {
        repository.save(listOf(TestElement(0), TestElement(1), TestElement(2)).iterator())

        listOf(0, 1, 2).forEach { id ->
            assertThat(innerRepo.contains(id), `is`(true))
            verify(policy, times(1)).notifyCreated(id)
        }
    }

    @Test
    fun `repository should report working if the elements are properly initialized`() {
        assertTrue(repository.isReady)
    }
}
